### PR TITLE
save device type instead of caching it

### PR DESCRIPTION
**Pull request by rawdaGastan**
### Issues
 - https://github.com/threefoldtech/zos/issues/1970

### CODE REVIEW

File: pkg/storage/filesystem/device.go
Comment (muhamadazmy) [inline pkg/storage/filesystem/device.go]: this can be any error type, but only `NotFound` is false, hence this method actually should return `(type, bool, error)

Comment (muhamadazmy) [reply pkg/storage/filesystem/device.go]: Consider that any error is a `false` can hide other problems with the disk or the file system where in matter of fact the disk should be considered `broken` or in a wrong state

File: pkg/storage/storage.go
Comment (muhamadazmy) [inline pkg/storage/storage.go]: this type then need to be stored to disk. So both cases if you get type by detection or from cache has to be written to disk.

**Review by muhamadazmy — CHANGES_REQUESTED** [review]
I have few comments. but also i personally don't deep nesting like that, so the code might get some re-organized and or moved to separate functions.

for example
```go

typ, err := s.diskType(device)

// where we have 
func diskType(device)  {
 //get type 
 if found {
   return
} 
 
 cached, found := cache.get()
 if !found {
   cached = detected() 
  }

// store on device
return
}
 
```

File: pkg/storage/filesystem/device.go
Comment (muhamadazmy) [inline pkg/storage/filesystem/device.go]: The correct way to do this is
```suggestion
	if os.IsNotExist(err) {
```

File: pkg/storage/storage.go
Comment (muhamadazmy) [inline pkg/storage/storage.go]: don't return here because this type need to be stored on the disk, if u check my psudo code, I check if NOT cache, i do detect then ALWAYS store the result on disk

File: pkg/storage/storage.go
Comment (muhamadazmy) [inline pkg/storage/storage.go]: I think this check can be moved to the top of this function no need to store the type because we force ssd ALWAYS except for vdd and vde which are always HDD

File: pkg/storage/storage.go
Comment (muhamadazmy) [inline pkg/storage/storage.go:249]: you need to continue after logging the error. (might also be a good idea to add them to broken pools)

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

**Review by muhamadazmy — DISMISSED** [review]
Please don't merge unless code is fully tested

File: pkg/storage/filesystem/device.go
Comment (xmonader) [inline pkg/storage/filesystem/device.go:71]: First time for me to know about the seektime tool, maybe we can enhance it to include that it's an approximation too https://github.com/maxux/seektime not 100% correct

File: pkg/storage/filesystem/device.go
Comment (xmonader) [inline pkg/storage/filesystem/device.go]: Why don't we use zos.DeviceType? I'm not sure why we created that alias declaration DeviceType = DeviceType?
Also the rest of the functions DetectType, Type are working with zos.DeviceType

File: pkg/storage/filesystem/device.go
Comment (xmonader) [inline pkg/storage/filesystem/device.go]: please include the full path as well so `/mnt/d.Name(), ...`

File: pkg/storage/filesystem/device.go
Comment (xmonader) [inline pkg/storage/filesystem/device.go]: let's extract the first argument of WriteFile into a varilable to improve the readability

File: pkg/storage/filesystem/device.go
Comment (xmonader) [inline pkg/storage/filesystem/device.go]: let's expand the docstring, that's based on `.seektime` file existing in the /mnt/DeviceName.seektime, that contains the device type being SSD or HDD maybe

File: pkg/storage/filesystem/device.go
Comment (xmonader) [inline pkg/storage/filesystem/device.go]: - let's extract to its own variable
- let's rename data to disktype

File: pkg/storage/filesystem/device.go
Comment (xmonader) [reply pkg/storage/filesystem/device.go]: also i'm not sure of the behavior of the return type device, bool, err

File: pkg/storage/filesystem/device.go
Comment (xmonader) [reply pkg/storage/filesystem/device.go]: actually I believe we shouldn't use `.seektime` extension, as it denotes the results of seektime (rpm/seektime) instead of actual harddisk type string, however if it's considered as "the output of seektime" tool, I'm good with that

**Comment by xmonader** [comment]
I noticed this in the seektime code https://github.com/maxux/seektime/issues/2 , not sure how it affects the accuracy or the calculations @maxux @muhamadazmy ?

**Comment by maxux** [comment]
Regarding https://github.com/maxux/seektime/issues/2, this doesn't affect any accuracy in our case.

But this https://github.com/threefoldtech/seektime/pull/3 was ready for a long time now and were never merged, but this could provide a way better help to troubleshoot in case of issue.

Does merging this PR would affect something in zos ? The json reply would change, maybe you need to support it ?

This pull requests will print `min` and `max` seektime, this helped me previously to diagnostic a bug where HDD were sleeping and disks needed to be woke up on the first check; `max` was set to few of seconds and `min` were real value, this helped to understand what happened. This should be merged IMO. This can aswell add some precision on the check and is always helpful.

**Comment by muhamadazmy** [comment]
@maxux current seektime version in zos return this

```
seektime -j /dev/vda
{"device": "/dev/vda", "type": "HDD", "elapsed": 564}
```
if the format is going to change then zos also need to have some modification to support his.

Right now it seems that seektime is only provided by the base image. so after merging of your branch, and if zos image is rebuilt, then zos will break. Instead what we need to do is provide a runtime package update for seektime so both the seektime version and the code to handle it to be installed in one go.

Can you please provide a `seektime` pre-build in the seektime release. A static build is the best or in worst case make sure it uses the same libc version of the base zos image.

Once this build is available, an flist can be built for seektime that can be included in zos

**Review by muhamadazmy — APPROVED** [review]

**Pull request merged**
